### PR TITLE
[libfuzzer] Initial attempt at supporting xpcrt

### PR DESCRIPTION
**Pull request by pyoor**
Adds setup and fuzzing paths for xpcrt targets.

Depends on https://github.com/MozillaSecurity/fuzzfetch/pull/93

### CODE REVIEW

**Comment by pyoor** [comment]
This no longer depends on https://github.com/MozillaSecurity/fuzzfetch/pull/93 since we're using `firefox -xpcshell`.

File: services/libfuzzer/libfuzzer.sh [hunk 1]
```diff
@@ -52,6 +52,37 @@ then
...
 fi
 
+if [[ -n "$XPCRT" ]]
+then
+  if [[ ! -e ~/.ssh/id_rsa.domino ]] || [[ ! -e ~/.ssh/id_rsa.domino-xpcshell ]]
+  then
+    targets=( "domino" "domino-xpcshell" )
+    for target in "${targets[@]}"
+    do
+      echo "$(get-tc-secret deploy-$target)"
```
Comment (jschwartzentruber) [inline services/libfuzzer/libfuzzer.sh:62]: don't print the key here, pls

File: services/libfuzzer/libfuzzer.sh [hunk 2]
```diff
@@ -52,6 +52,37 @@ then
...
+      retry get-tc-secret "deploy-$target" "$HOME/.ssh/id_rsa.${target}"
+      chmod 0600 "$HOME/.ssh/id_rsa.${target}"
```
Comment (jschwartzentruber) [inline services/libfuzzer/libfuzzer.sh:64]: retry/chmod are handled by `get-tc-secret`

File: services/libfuzzer/libfuzzer.sh [hunk 3]
```diff
@@ -52,6 +52,37 @@ then
...
+      cat >> ~/.ssh/config <<-EOF
+			Host $target
+			Hostname github.com
+			IdentityFile ~/.ssh/id_rsa.$target
+			EOF
+    done
+  fi
+
+  npm set //registry.npmjs.org/:_authToken="$(get-tc-secret deploy-npm)"
```
Comment (jschwartzentruber) [inline services/libfuzzer/libfuzzer.sh:73]: you'll want `set +x` and `set -x` wrapped around this line

**Review by jschwartzentruber — CHANGES_REQUESTED** [review]

**Review by choller — APPROVED** [review]
Please also make sure that existing libFuzzer targets won't break on this by giving them a test run.

File: services/libfuzzer/libfuzzer.sh [hunk 4]
```diff
@@ -73,7 +109,12 @@ TARGET_BIN="$(./setup-target.sh)"
 
 FUZZDATA_URL="https://github.com/mozillasecurity/fuzzdata.git/trunk"
 function run-afl-libfuzzer-daemon () {
-  python3 ./fuzzmanager/misc/afl-libfuzzer/afl-libfuzzer-daemon.py "$@"
+  if [[ -n "$XPCRT" ]]
+  then
+    xvfb-run python3 ./fuzzmanager/misc/afl-libfuzzer/afl-libfuzzer-daemon.py "$@"
```
[the remark this answers is not in the document]
Comment (pyoor) [reply services/libfuzzer/libfuzzer.sh:114]: It appears that some DOM interfaces exposed in xpcshell attempt to access gtk during initialization.  The following is the result of `new DOMParser().parseFromString()`

```
[Job 0] ==464== ERROR: libFuzzer: fuzz target exited
[Job 0]     #0 0x56502415d391 in __sanitizer_print_stack_trace /builds/worker/fetches/llvm-project/llvm/projects/compiler-rt/lib/asan/asan_stack.cpp:86:3
[Job 0]     #1 0x565024368b21 in fuzzer::PrintStackTrace() /builds/worker/checkouts/gecko/tools/fuzzing/libfuzzer/FuzzerUtil.cpp:210:5
[Job 0]     #2 0x56502434fe09 in fuzzer::Fuzzer::ExitCallback() /builds/worker/checkouts/gecko/tools/fuzzing/libfuzzer/FuzzerLoop.cpp:250:3
[Job 0]     #3 0x7fd84801ba26  (/lib/x86_64-linux-gnu/libc.so.6+0x49a26)
[Job 0]     #4 0x7fd84801bbdf in exit (/lib/x86_64-linux-gnu/libc.so.6+0x49bdf)
[Job 0]     #5 0x7fd843cb905a in gtk_init (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x24e05a)
[Job 0]     #6 0x7fd8291d7591 in gfxPlatformGtk::gfxPlatformGtk() /builds/worker/checkouts/gecko/gfx/thebes/gfxPlatformGtk.cpp:95:5
[Job 0]     #7 0x7fd8291be21b in gfxPlatform::Init() /builds/worker/checkouts/gecko/gfx/thebes/gfxPlatform.cpp:913:19
[Job 0]     #8 0x7fd8291c2660 in GetPlatform /builds/worker/checkouts/gecko/gfx/thebes/gfxPlatform.cpp:481:5
[Job 0]     #9 0x7fd8291c2660 in gfxPlatform::InitializeCMS() /builds/worker/checkouts/gecko/gfx/thebes/gfxPlatform.cpp:2170:9
[Job 0]     #10 0x7fd82de56433 in GetCMSMode /builds/worker/workspace/obj-build/dist/include/gfxPlatform.h:543:5
[Job 0]     #11 0x7fd82de56433 in nsXPLookAndFeel::GetColorValue(mozilla::StyleSystemColor, mozilla::LookAndFeel::ColorScheme, mozilla::LookAndFeel::UseStandins, unsigned int&) /builds/worker/checkouts/gecko/widget/nsXPLookAndFeel.cpp:711:9
[Job 0]     #12 0x7fd82de599be in mozilla::LookAndFeel::GetColor(mozilla::StyleSystemColor, mozilla::LookAndFeel::ColorScheme, mozilla::LookAndFeel::UseStandins) /builds/worker/checkouts/gecko/widget/nsXPLookAndFeel.cpp:974:47
[Job 0]     #13 0x7fd82de00b7b in Color /builds/worker/workspace/obj-build/dist/include/mozilla/LookAndFeel.h:467:12
[Job 0]     #14 0x7fd82de00b7b in ThemedAccentColor /builds/worker/checkouts/gecko/widget/nsNativeBasicTheme.cpp:147:37
[Job 0]     #15 0x7fd82de00b7b in nsNativeBasicTheme::RecomputeAccentColors() /builds/worker/checkouts/gecko/widget/nsNativeBasicTheme.cpp:259:20
[Job 0]     #16 0x7fd82de00976 in LookAndFeelChanged /builds/worker/checkouts/gecko/widget/nsNativeBasicTheme.cpp:246:3
[Job 0]     #17 0x7fd82de00976 in nsNativeBasicTheme::Init() /builds/worker/checkouts/gecko/widget/nsNativeBasicTheme.cpp:236:3
[Job 0]     #18 0x7fd82de54ec2 in nsXPLookAndFeel::GetInstance() /builds/worker/checkouts/gecko/widget/nsXPLookAndFeel.cpp:356:3
[Job 0]     #19 0x7fd82de59e20 in mozilla::LookAndFeel::GetInt(mozilla::LookAndFeel::IntID, int*) /builds/worker/checkouts/gecko/widget/nsXPLookAndFeel.cpp:1038:10
[Job 0]     #20 0x7fd82e20babf in GetInt /builds/worker/workspace/obj-build/dist/include/mozilla/LookAndFeel.h:493:9
[Job 0]     #21 0x7fd82e20babf in UseAccessibilityTheme /builds/worker/checkouts/gecko/layout/style/PreferenceSheet.cpp:66:12
[Job 0]     #22 0x7fd82e20babf in mozilla::PreferenceSheet::Prefs::Load(bool) /builds/worker/checkouts/gecko/layout/style/PreferenceSheet.cpp:84:28
[Job 0]     #23 0x7fd82e20cf30 in mozilla::PreferenceSheet::Initialize() /builds/worker/checkouts/gecko/layout/style/PreferenceSheet.cpp:149:17
[Job 0]     #24 0x7fd82e213143 in EnsureInitialized /builds/worker/checkouts/gecko/layout/style/PreferenceSheet.h:59:5
[Job 0]     #25 0x7fd82e213143 in mozilla::ServoStyleSet::ServoStyleSet(mozilla::dom::Document&) /builds/worker/checkouts/gecko/layout/style/ServoStyleSet.cpp:115:3
[Job 0]     #26 0x7fd829958067 in mozilla::detail::UniqueSelector<mozilla::ServoStyleSet>::SingleObject mozilla::MakeUnique<mozilla::ServoStyleSet, mozilla::dom::Document&>(mozilla::dom::Document&) /builds/worker/workspace/obj-build/dist/include/mozilla/UniquePtr.h:609:27
[Job 0]     #27 0x7fd829957c5f in mozilla::dom::Document::Init() /builds/worker/checkouts/gecko/dom/base/Document.cpp:2646:15
[Job 0]     #28 0x7fd82c06486d in nsHTMLDocument::Init() /builds/worker/checkouts/gecko/dom/html/nsHTMLDocument.cpp:145:27
[Job 0]     #29 0x7fd82c064278 in NS_NewHTMLDocument(mozilla::dom::Document**, bool) /builds/worker/checkouts/gecko/dom/html/nsHTMLDocument.cpp:111:22
[Job 0]     #30 0x7fd82d81bf5d in NS_NewDOMDocument(mozilla::dom::Document**, nsTSubstring<char16_t> const&, nsTSubstring<char16_t> const&, mozilla::dom::DocumentType*, nsIURI*, nsIURI*, nsIPrincipal*, bool, nsIGlobalObject*, DocumentFlavor) /builds/worker/checkouts/gecko/dom/xml/XMLDocument.cpp:72:10
[Job 0]     #31 0x7fd8298cfab0 in mozilla::dom::DOMParser::SetUpDocument(DocumentFlavor, mozilla::ErrorResult&) /builds/worker/checkouts/gecko/dom/base/DOMParser.cpp:314:17
[Job 0]     #32 0x7fd8298cf2e4 in mozilla::dom::DOMParser::ParseFromString(nsTSubstring<char16_t> const&, mozilla::dom::SupportedType, mozilla::ErrorResult&) /builds/worker/checkouts/gecko/dom/base/DOMParser.cpp:62:35
```

**Comment by pyoor** [comment]
> Please also make sure that existing libFuzzer targets won't break on this by giving them a test run.

I tested these changes against the SDP fuzzer and it seems to work as expected.

File: services/libfuzzer/libfuzzer.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (pyoor) [reply services/libfuzzer/libfuzzer.sh]: I've replaced `xvfb-run` with `MOZ_HEADLESS=1` in 5460700.

**Pull request merged**